### ISSUE TITLE
Retry failed uploads in background (Pictures & Videos)

### ISSUE THREAD

**Issue by enoch85** [nextcloud/ios#306]
In https://github.com/nextcloud/ios/issues/215#issuecomment-312690578 we had a discussion about uploading pictures and videos in the background. Right now it doesn't and you have to open the app manually to retry every 10 seconds in foreground. 

This should stay as is, but it would be good if we retried failed uploads in the background as well. For example let's say you were on holiday, you took loads of pictures and videos that were uploaded sporadically due to bad WIFI or cellular, battery died (or whatever). When you come home you expect all the pictures and videos are uploaded to Nextcloud, but they are not. Several pictures are missing, heck! What would solve this? Automatic Retry of Failed Uploads.

**TL;DR**
- I know that uploading stuff in the background is resource intensive especially videos. That's why there should be a switch; `Retry Failed Uploads in the Background.` 
- Retry every 6 hours or so because of above if activated.
- Check pending uploads after battery died, or the app crashed. Keep unsuccessful uploads meta in cache maybe?
- Retry every time there is a significant change in location (that's how it's currently working with Apple)
- ownCloud does this in the background, why can't we?
- Maybe introduce a list with all the failed uploads so that you can manually retry as well? Or even better, use the current `Transfers` view.

@nextcloud/ios What do you think?

**Comment by roedie** [comment]
I would love to see this as well. Since I do not open the app regularly on the iPhone, retrying in the background seems like the way to go.

**Comment by enoch85** [comment]
@marinofaggiana Could we have this in 2.17.5?

**Comment by marinofaggiana** [comment]
Next week

**Comment by marinofaggiana** [comment]
retry available, 2.17.5

**Comment by rakekniven** [comment]
@enoch85 Still open? v2.17.5 is out. Please verify.

**Comment by enoch85** [comment]
@rakekniven Nope, still not fixed. Went out on a trip just to test. Now not even photos are uploaded and videos get uploaded wiht 0 byte.

![screenshot from 2017-07-23 15 56 34](https://user-images.githubusercontent.com/4511254/28499851-8bd54b04-6fbf-11e7-8f4f-f02701ff2a83.png)

**Current settings are:**

![17-07-23 15-57-54 5278](https://user-images.githubusercontent.com/4511254/28499881-0db69416-6fc0-11e7-8d63-25b432a370fe.png)

**Activity feed:**

![17-07-23 15-58-55 5279](https://user-images.githubusercontent.com/4511254/28499884-15be2502-6fc0-11e7-82f9-ff57cb4a55a0.png)

Nådde en maxtidsgräns = Reached max timeout limit

If I open the app in fourground it works and videos get uploaded as they should, but that's not the desired function here.

**Comment by marinofaggiana** [comment]
mmmm mumble mumble .... very strange .... 0 length ...

**Comment by enoch85** [comment]
> 0 length 

That was yesterday, and the videos were not uploaded at all in background but today they got uploaded in the foreground with 0 byte.

**Comment by marinofaggiana** [comment]
I am not sure of this issue @enoch85  :-) When u have time make a new with 2.17.5 (not from previous version)

**Comment by enoch85** [comment]
As I said:

> Nope, still not fixed. Went out on a trip just to test. Now not even photos are uploaded and videos get uploaded with 0 byte.

I went on a trip, took 2 videos and 3 photos. Went home and none of them were uploaded. Not even when reaching WIFI so this is broken. Don't mind the 0 byte stuff, fact is that videos are **not** uploaded, and as you can see I got new errors with 2.17.5 that weren't there yesterday.

**Comment by enoch85** [comment]
@marinofaggiana If you send me your email address I can create an account for you on my server so that you can test yourself. Thanks!

**Comment by marinofaggiana** [comment]
Thanks but the issue is not the server ... (I think) the errors says : server unreachable ... but where are 3 photo and 2 video errors ?

**Comment by marinofaggiana** [comment]
and why error autouploadall ??

**Comment by enoch85** [comment]
> and why error autouploadall ??

I don't know, you're the dev :)

You could take a look at ownCloud iOS code and see how they do to get inspired.

**Comment by enoch85** [comment]
What does the option "Upload whole camera roll" do? Does it make any difference?

**Comment by marinofaggiana** [comment]
hahaha the calls of API are the same ... ;) autouploadall is a error for select Auto Upload all Photo switch ON but your phone was offline (server  unreachable) ... I don't view errors for your 3 photo and 2 videos

**Comment by marinofaggiana** [comment]
ohhh YES !! this option upload ALL camera roll

**Comment by enoch85** [comment]
> I don't view errors for your 3 photo and 2 videos

I can't fit them in the screen. And now I deleted the log to get a fresh one. 

But this issue is about retrying failed uploads. so if they failed, they should be retried which they don't. Not even when I come home to my wifi.

> ohhh YES !! this option upload ALL camera roll

What do you mean? I asked what the function was, but you didn't answer?

**Comment by marinofaggiana** [comment]
yes I have answer, this function upload all camera roll not only the last detect

**Comment by enoch85** [comment]
> yes I have answer, this function upload all camera roll not only the last detect

So what's the difference in a Auto Upload situation?

`Upload whole camera roll`: **ON**:
- It uploads all photos manually everytime I open the app in foreground?
- It Auto Uploads all photos on camera roll twice (one time when photo is taken, another time when I open the app in foreground)?

`Upload whole camera roll`: **OFF**:
- Auto Upload doesn't work since it doens't upload all the photos and videos on the camera roll.

Please be precise, I don't understand when you write "ohhh YES !! this option upload ALL camera roll"

Thank you.

**Comment by rakekniven** [comment]
Hello all, sitting on my veranda and following your conversation.

Are there any testcases I can do?

Use the app daily and background upload is mystic for me. My habit is to open nextcloud app to have photos uploaded.

Enjoy sunday! Whatever you are doin' :-)

**Comment by rakekniven** [comment]
Wow, get away from your keyboard. Put your feet in cool water and get a drink.  🍹

**Comment by enoch85** [comment]
@rakekniven 

> My habit is to open nextcloud app to have photos uploaded.

https://github.com/nextcloud/ios/issues/215#issuecomment-307195988

So, what you could do is to:
1. Close the app completely, double tap home button and remove it from background apps as well. 
2. Take some photos and videos
3. Move around so that the GPS get triggered.
4. See if the photos and videos are uploaded (without opening the app, check on the server directly)

The point here is to upload photos without even opening the app, and retry in the background automatically if it failed the first time.

@marinofaggiana You still din't answer my question: https://github.com/nextcloud/ios/issues/306#issuecomment-317258238

**Comment by enoch85** [comment]
@marinofaggiana 

> hahaha the calls of API are the same ... ;)

Yeah, but then it's strange that it works on ownCloud but not here...

**Comment by rakekniven** [comment]
@enoch85 Will be on holiday in a few days and my GPS will trigger app. That's for sure 😆 

Will follow your testcase and understood basic function.

**Comment by marinofaggiana** [comment]
boys I close this, reopen if exists news

**Comment by enoch85** [comment]
> boys I close this, reopen if exists news

What?! It doesn't work. Then why closing?

**Comment by marinofaggiana** [comment]
Please more detail, from several users.

**Comment by enoch85** [comment]
@marinofaggiana Please give me your email adress to that I can write to you personally instead of discussing publicly. Or join #nextcloud on IRC.

Thanks!

**Comment by enoch85** [comment]
Ping @roedie as well for more user input.

**Comment by enoch85** [comment]
@rakekniven Did you have any chance to test this yet?

**Comment by roedie** [comment]
I've tested this week, but it seems to upload everything.

**Comment by enoch85** [comment]
@roedie You are only on WIFI right? And what about Videos?

**Comment by roedie** [comment]
@enoch85 I'm only on WiFi indeed. Videos upload just as well.

**Comment by enoch85** [comment]
OK, have you ever had a failed upload? Can you try to make an upload fail, e.g. go from wifi to cellular while a upload is happening or something like that?

**Comment by enoch85** [comment]
Also, do you have the "upload all files on camera roll" option enabled?

**Comment by roedie** [comment]
I'll have a try at creating a failed upload.

I do not have the "upload all files on camera roll" option enabled since it will re-upload everything I already have.

**Comment by enoch85** [comment]
Thanks @roedie!

**Comment by enoch85** [comment]
I changed to Android instead so I can't test. But I still think that everything in my first post should be implemented. 

What's the status here?
